# Web Scraping Project Solutions

In [2]:
import pandas as pd
from bs4 import BeautifulSoup
import requests

#### Q1. Write a python program to display all the header tags from ‘en.wikipedia.org/wiki/Main_Page’.

In [2]:
def header_tags(url):
    a = requests.get(url)
    src = a.content
    soup = BeautifulSoup(src,'html.parser')
    return soup.find_all(["h1","h2","h3","h4","h5"])

# Calling Function
header_tags('https://en.wikipedia.org/wiki/Main_Page')

[<h1 class="firstHeading" id="firstHeading">Main Page</h1>,
 <h2 class="mp-h2" id="mp-tfa-h2"><span id="From_today.27s_featured_article"></span><span class="mw-headline" id="From_today's_featured_article">From today's featured article</span></h2>,
 <h2 class="mp-h2" id="mp-dyk-h2"><span class="mw-headline" id="Did_you_know_...">Did you know ...</span></h2>,
 <h2 class="mp-h2" id="mp-itn-h2"><span class="mw-headline" id="In_the_news">In the news</span></h2>,
 <h2 class="mp-h2" id="mp-otd-h2"><span class="mw-headline" id="On_this_day">On this day</span></h2>,
 <h2 class="mp-h2" id="mp-tfl-h2"><span id="From_today.27s_featured_list"></span><span class="mw-headline" id="From_today's_featured_list">From today's featured list</span></h2>,
 <h2 class="mp-h2" id="mp-tfp-h2"><span id="Today.27s_featured_picture"></span><span class="mw-headline" id="Today's_featured_picture">Today's featured picture</span></h2>,
 <h2 class="mp-h2" id="mp-other"><span class="mw-headline" id="Other_areas_of_Wikipe

#### Q2. Write a python program to display IMDB’s Top rated 100 movies’ data (i.e. Name, IMDB rating, Year of release) and make a DataFrame

In [3]:
def IMDB_top_100(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.content, 'html.parser')
    names = []
    ratings=[]
    years = []
    for i in soup.find_all('td', class_ = 'titleColumn'):
        names.append(i.text.replace('\n', "")[14:-6])
    for i in soup.find_all('td', class_ = 'ratingColumn imdbRating'):
        ratings.append(float(i.text.replace('\n', "")))
    for i in soup.find_all('span', class_='secondaryInfo'):
        years.append(int(i.text[1:-1]))
    df = pd.DataFrame({'Names':names[:100], 'Ratings':ratings[:100], 'Years of release':years[:100]})
    return df
#calling function
IMDB_top_100('https://www.imdb.com/chart/top/')

Names  Ratings  Years of release
0                 The Shawshank Redemption(1994)      9.2              1994
1                            The Godfather(1972)      9.1              1972
2                   The Godfather: Part II(1974)      9.0              1974
3                          The Dark Knight(2008)      9.0              2008
4                             12 Angry Men(1957)      8.9              1957
..                                           ...      ...               ...
95                     Singin' in the Rain(1952)      8.3              1952
96                      North by Northwest(1959)      8.3              1959
97   Eternal Sunshine of the Spotless Mind(2004)      8.3              2004
98                            Idi i smotri(1985)      8.3              1985
99                     Ladri di biciclette(1948)      8.3              1948

[100 rows x 3 columns]

#### Q3. Write a python program to display IMDB’s Top rated 100 Indian movies’ data (i.e. Name, IMDB rating, Year of release) and make a DataFrame. 

In [4]:
def Indian_IMDB_top_100(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.content, 'html.parser')
    names = []
    ratings=[]
    years = []
    for i in soup.find_all('td', class_ = 'titleColumn'):
        names.append(i.text.replace('\n', "")[14:-6])
    for i in soup.find_all('td', class_ = 'ratingColumn imdbRating'):
        ratings.append(float(i.text.replace('\n', "")))
    for i in soup.find_all('span', class_='secondaryInfo'):
        years.append(int(i.text[1:-1]))
    df = pd.DataFrame({'Names':names[:100], 'Ratings':ratings[:100], 'Years of release':years[:100]})
    return df
Indian_IMDB_top_100('https://www.imdb.com/india/top-rated-indian-movies/?pf_rd_m=A2FGELUUNOQJNL&pf_rd_p=8a7876cd-2844-4017-846a-2c0876945b7b&pf_rd_r=ZHXFB84HZG5YP7EWJ71C&pf_rd_s=right-5&pf_rd_t=15506&pf_rd_i=top&ref_=chttp_india_tr_rhs_1')

Names  Ratings  Years of release
0             Pather Panchali      8.5              1955
1                     Nayakan      8.5              1987
2                  Anbe Sivam      8.5              2003
3           Pariyerum Perumal      8.5              2018
4                  Drishyam 2      8.5              2021
..                        ...      ...               ...
95   Uri: The Surgical Strike      8.1              2018
96                 Chhichhore      8.1              2019
97                  Virumandi      8.1              2004
98                         PK      8.1              2014
99                      Iqbal      8.1              2005

[100 rows x 3 columns]

#### Q4. Write a python program to scrap book name, author name, genre and book review of any 5 books from ‘www.bookpage.com’.

In [5]:
def bookpage():
    r = requests.get('https://bookpage.com/reviews')
    soup = BeautifulSoup(r.content, 'html.parser')
    urls=[]
    url_tags = soup.find_all('div', class_='row-fluid article-row')
    for i in url_tags[0:5]:
        urls.append(i.find('h4').find_all('a')[0]['href'])
    bookname = []
    authorname = []
    genre = []
    bookreview = []
    for url in urls:
        r = requests.get('https://bookpage.com'+url)
        soup = BeautifulSoup(r.content,'html.parser')
        bookname.append(soup.find('h1', class_='italic').text.replace('\n',""))
        authorname.append(soup.find('h4', class_='sans').text.replace('\n',""))
        genre.append(soup.find('p', class_='genre-links').text.replace('\n',""))
        bookreview.append(soup.find('div', class_='article-body').text.replace('\n',""))
    df = pd.DataFrame({'Book names':bookname,
                       'Author names':authorname,
                       'Genres':genre,
                       'Book reviews':bookreview })
    return df
bookpage()

Book names        Author names  \
0          You People      Nikita Lalwani   
1  ★ Ship in a Bottle       Andrew Prahin   
2    Tokyo Ever After          Emiko Jean   
3        The Kingdoms      Natasha Pulley   
4         New Yorkers  Craig Taylor, Cast   

                                 Genres  \
0            Fiction / Literary Fiction   
1  Children's / Children's Picture Book   
2                       YA / YA Fiction   
3         Fiction / Speculative Fiction   
4          Audio / Nonfiction / Culture   

                                        Book reviews  
0  The incident that gives You People its title o...  
1  Poor Mouse. She lives with Cat, and needless t...  
2  There’s no shortage of YA novels in which a co...  
3  When Joe Tournier steps off a train from Glasg...  
4  “Part of loving New York is just mourning the ...

#### Q5. Write a python program to scrape cricket rankings from ‘www.icc-cricket.com’. You have to scrape:
i) Top 10 ODI teams in men’s cricket along with the records for matches, points and rating.                                 
ii) Top 10 ODI Batsmen in men along with the records of their team and rating.                                             
iii) Top 10 ODI bowlers along with the records of their team and rating.

#### Solution:
i) Top 10 ODI teams in men’s cricket along with the records for matches, points and rating.

In [19]:
def Top_10_ODI_team(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.content,'html.parser')
    teams = []
    matches = []
    points = []
    ratings = []
    for i in soup.find_all('span',attrs={'u-hide-phablet'}):
        teams.append(i.text)
    matches.append(soup.find('td',class_='rankings-block__banner--matches').text)
    points.append(soup.find('td',class_='rankings-block__banner--points').text)
    ratings.append(soup.find('td', class_='rankings-block__banner--rating u-text-right').text.replace('\n',"").strip())
    count=0
    for i in list(soup.find_all('td',class_='table-body__cell u-center-text'))[:18]:
        if count%2==0:
            matches.append(i.text)
        else:
            points.append(i.text)
        count+=1
    for i in soup.find_all('td', class_='table-body__cell u-text-right rating'):
        ratings.append(int(i.text.replace('\n',"").strip()))
    df = pd.DataFrame({'Teams': teams[:10], 
                       'Matches':matches,
                       'Points': points,
                       'Ratings': ratings[:10]})
    return df
Top_10_ODI_team('https://www.icc-cricket.com/rankings/mens/team-rankings/odi')

Teams Matches Points Ratings
0   New Zealand      17  2,054     121
1     Australia      25  2,945     118
2         India      29  3,344     115
3       England      27  3,100     115
4  South Africa      20  2,137     107
5      Pakistan      24  2,323      97
6    Bangladesh      26  2,413      93
7   West Indies      27  2,222      82
8     Sri Lanka      23  1,733      75
9   Afghanistan      17  1,054      62

ii) Top 10 ODI Batsmen in men along with the records of their team and rating.

In [22]:
def Top_10_ODI_batsman(url):
    r=requests.get(url)
    soup=BeautifulSoup(r.content,'html.parser')
    name=[]
    team=[]
    point=[]
    name.append(soup.find('div',class_="rankings-block__banner--name-large").text.replace('\n',""))
    for i in list(soup.find_all('td',class_='table-body__cell rankings-table__name name'))[:9]:
        name.append(i.text.replace('\n',''))
    team.append(soup.find('div',class_="rankings-block__banner--nationality").text.replace('\n',""))
    for i in list(soup.find_all('td',class_='table-body__cell nationality-logo rankings-table__team'))[:9]:
        team.append(i.text.replace('\n',''))
    point.append(soup.find('div',class_="rankings-block__banner--rating").text)
    for i in list(soup.find_all('td',class_='table-body__cell rating'))[:9]:
        point.append(i.text)
    df=pd.DataFrame({'Name': name,
                     'Team':team,
                     'points':point})
    return df


# Calling Function
Top_10_ODI_batsman("https://www.icc-cricket.com/rankings/mens/player-rankings/odi/batting")

Name Team points
0           Babar Azam  PAK    865
1          Virat Kohli  IND    857
2         Rohit Sharma  IND    825
3          Ross Taylor   NZ    801
4          Aaron Finch  AUS    791
5       Jonny Bairstow  ENG    785
6         Fakhar Zaman  PAK    778
7  Francois du Plessis   SA    778
8         David Warner  AUS    773
9            Shai Hope   WI    773

iii) Top 10 ODI bowlers along with the records of their team and rating.

In [23]:
def Top_10_ODI_bowler(url):
    r=requests.get(url)
    soup=BeautifulSoup(r.content,'html.parser')
    name=[]
    team=[]
    point=[]
    name.append(soup.find('div',attrs={"rankings-block__banner--name-large"}).text)
    for i in list(soup.find_all('td',attrs={'table-body__cell rankings-table__name name'}))[:9]:
        name.append(i.text.replace('\n',''))
    team.append(soup.find('div',attrs={"rankings-block__banner--nationality"}).text.replace('\n',""))
    for i in list(soup.find_all('td',attrs={'table-body__cell nationality-logo rankings-table__team'}))[:9]:
        team.append(i.text.replace('\n',''))
    point.append(soup.find('div',attrs={"rankings-block__banner--rating"}).text)
    for i in list(soup.find_all('td',attrs={'table-body__cell rating'}))[:9]:
        point.append(i.text)
    df=pd.DataFrame({'Name': name,
                     'Team':team,
                     'points':point})
    return df

# Calling Function
Top_10_ODI_batsman("https://www.icc-cricket.com/rankings/mens/player-rankings/odi/bowling")

Name Team points
0        Trent Boult   NZ    737
1       Mehedi Hasan  BAN    725
2   Mujeeb Ur Rahman  AFG    708
3         Matt Henry   NZ    691
4     Jasprit Bumrah  IND    690
5      Kagiso Rabada   SA    666
6       Chris Woakes  ENG    665
7     Josh Hazlewood  AUS    660
8  Mustafizur Rahman  BAN    652
9        Pat Cummins  AUS    646

#### Q6. Write a python program to scrape cricket rankings from ‘www.icc-cricket.com’. You have to scrape:
i) Top 10 ODI teams in women’s cricket along with the records for matches, points and rating.                                  
ii) Top 10 women’s ODI players along with the records of their team and rating.                                                
iii) Top 10 women’s ODI all-rounder along with the records of their team and rating. 

#### Solutions:
i) Top 10 ODI teams in women’s cricket along with the records for matches, points and rating.

In [24]:
def Top_10_Women_ODI_team(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.content,'html.parser')
    teams = []
    matches = []
    points = []
    ratings = []
    for i in soup.find_all('span',class_='u-hide-phablet'):
        teams.append(i.text)
    matches.append(soup.find('td',class_='rankings-block__banner--matches').text)
    points.append(soup.find('td',class_='rankings-block__banner--points').text)
    ratings.append(soup.find('td', class_='rankings-block__banner--rating u-text-right').text.replace('\n',"").strip())
    count=0
    for i in list(soup.find_all('td',class_='table-body__cell u-center-text'))[:18]:
        if count%2==0:
            matches.append(i.text)
        else:
            points.append(i.text)
        count+=1
    for i in soup.find_all('td', class_='table-body__cell u-text-right rating'):
        ratings.append(int(i.text.replace('\n',"").strip()))
    df = pd.DataFrame({'Teams': teams[:10], 
                       'Matches':matches,
                       'Points': points,
                       'Ratings': ratings[:10]})
    return df
Top_10_Women_ODI_team('https://www.icc-cricket.com/rankings/womens/team-rankings/odi')
    

Teams Matches Points Ratings
0     Australia      18  2,955     164
1  South Africa      24  2,828     118
2       England      17  1,993     117
3         India      20  2,226     111
4   New Zealand      21  1,947      93
5   West Indies      12  1,025      85
6      Pakistan      15  1,101      73
7    Bangladesh       5    306      61
8     Sri Lanka      11    519      47
9       Ireland       2     25      13

ii) Top 10 women’s ODI players along with the records of their team and rating.

In [27]:
def Top_10_Women_ODI_player(url):
    r=requests.get(url)
    soup=BeautifulSoup(r.content,'html.parser')
    name=[]
    team=[]
    point=[]
    name.append(soup.find('div',class_="rankings-block__banner--name-large").text)
    for i in list(soup.find_all('td',class_='table-body__cell rankings-table__name name'))[:9]:
        name.append(i.text.replace('\n',''))
    team.append(soup.find('div',class_="rankings-block__banner--nationality").text.replace('\n', ""))
    for i in list(soup.find_all('td',class_='table-body__cell nationality-logo rankings-table__team'))[:9]:
        team.append(i.text.replace('\n',''))
    point.append(soup.find('div',class_="rankings-block__banner--rating").text)
    for i in list(soup.find_all('td',class_='table-body__cell rating'))[:9]:
        point.append(i.text)
    df=pd.DataFrame({'Name': name,
                     'Team':team,
                     'points':point})
    return df

# Calling Function
Top_10_Women_ODI_player("https://www.icc-cricket.com/rankings/womens/player-rankings/odi/batting")

Name Team points
0     Tammy Beaumont  ENG    765
1        Lizelle Lee   SA    758
2       Alyssa Healy  AUS    756
3    Stafanie Taylor   WI    746
4        Meg Lanning  AUS    723
5  Amy Satterthwaite   NZ    715
6    Smriti Mandhana  IND    710
7        Mithali Raj  IND    709
8     Natalie Sciver  ENG    685
9    Laura Wolvaardt   SA    683

iii) Top 10 women’s ODI all-rounder along with the records of their team and rating.

In [29]:
def Top_10_Women_ODI_Allrounder(url):
    r=requests.get(url)
    soup=BeautifulSoup(r.content,'html.parser')
    name=[]
    team=[]
    point=[]
    name.append(soup.find('div',class_="rankings-block__banner--name-large").text)
    for i in list(soup.find_all('td',class_='table-body__cell rankings-table__name name'))[:9]:
        name.append(i.text.replace('\n',''))
    team.append(soup.find('div',class_="rankings-block__banner--nationality").text.replace('\n',""))
    for i in list(soup.find_all('td',class_='table-body__cell nationality-logo rankings-table__team'))[:9]:
        team.append(i.text.replace('\n',''))
    point.append(soup.find('div',class_="rankings-block__banner--rating").text)
    for i in list(soup.find_all('td',class_='table-body__cell rating'))[:9]:
        point.append(i.text)
    df=pd.DataFrame({'Name': name,
                     'Team':team,
                     'points':point})
    return df

# Calling function
Top_10_Women_ODI_Allrounder("https://www.icc-cricket.com/rankings/womens/player-rankings/odi/all-rounder")

Name Team points
0    Marizanne Kapp   SA    418
1      Ellyse Perry  AUS    418
2   Stafanie Taylor   WI    410
3    Natalie Sciver  ENG    349
4     Deepti Sharma  IND    343
5     Jess Jonassen  AUS    307
6  Ashleigh Gardner  AUS    252
7  Dane van Niekerk   SA    243
8     Sophie Devine   NZ    242
9       Amelia Kerr   NZ    236

#### Q7. Write a python program to scrape details of all the mobile phones under Rs. 20,000 listed on Amazon.in. The scraped data should include Product Name, Price, Image URL and Average Rating.

In [10]:
r = requests.get('https://www.amazon.in/s?k=mobile+under+20000&rh=p_36%3A-2000000&qid=1622699513&rnid=1318502031&ref=sr_nr_p_36_4')
print(r.status_code)

503


The ratings of the Amazon.in cannot be scraped by normal method as we are getting status code as 503, Selenium is needed for it which is not yet taught to us. Hence skiping this question as told by Sajid Choudhary.

#### Q8. Write a python program to extract information about the local weather from the National Weather Service website of USA, https://www.weather.gov/ for the city, San Francisco. You need to extract data about 7 day extended forecast display for the city. The data should include period, short description, temperature and description.

In [20]:
def weather(url):
    r=requests.get(url)
    soup=BeautifulSoup(r.content,'html.parser')
    period=[]
    short_desc=[]
    temp=[]
    desc=[]
    for i in soup.find_all('p',class_='period-name'):
        period.append(i.text)
    for i in soup.find_all('p',class_='short-desc'):
        short_desc.append(i.text)
    for i in soup.find_all('p',class_='short-desc'):
        if i.next_sibling is not None:
            temp.append(i.next_sibling.text)
        else:
            temp.append(' ')
    for i in soup.find_all('div',class_="col-sm-10 forecast-text"):
        desc.append(i.text)
    df=pd.DataFrame({"Period":period,
                     "Short_description":short_desc,
                     "Temperature":temp,
                     "Description":desc[:9]})
    return df
weather('https://forecast.weather.gov/MapClick.php?lat=37.777120000000025&lon=-122.41963999999996#.YLc5FKgzYb4')

Period                         Short_description  Temperature  \
0          Today   Mostly Cloudythen MostlySunny andBreezy  High: 63 °F   
1        Tonight  Partly Cloudyand Breezythen MostlyCloudy   Low: 53 °F   
2       Thursday          Partly Sunnythen Sunnyand Breezy  High: 68 °F   
3  ThursdayNight   Mostly Clearand Breezythen PartlyCloudy   Low: 53 °F   
4         Friday          Mostly Sunnythen Sunnyand Breezy  High: 67 °F   
5    FridayNight    Mostly Clearand Breezythen MostlyClear   Low: 53 °F   
6       Saturday                                     Sunny  High: 68 °F   
7  SaturdayNight                              Mostly Clear   Low: 52 °F   
8         Sunday                                     Sunny  High: 67 °F   

                                         Description  
0  Mostly cloudy, then gradually becoming sunny, ...  
1  Mostly cloudy, with a low around 53. Breezy, w...  
2  Mostly cloudy, then gradually becoming sunny, ...  
3  Mostly clear, with a low around 53. Breezy, wi...  
4  Mostly sunny, with a high near 67. Breezy, wit...  
5       Mostly clear, with a low around 53. Breezy.   
6                        Sunny, with a high near 68.  
7                Mostly clear, with a low around 52.  
8                        Sunny, with a high near 67.

#### Q9. Write a python program to scrape fresher job listing from 'https://internshala.com/'. It should include job title, company name, CTC, and apply date.

In [92]:
def internshala_jobs(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.content, 'html.parser')
    job_title = []
    company = []
    lst = []
    CTC = []
    Apply_date = []

    for i in soup.find_all('div', class_='heading_4_5 profile'):
        job_title.append(i.text.replace('\n', ''))
    for i in soup.find_all('div', class_ = 'heading_6 company_name'):
        company.append(i.text.replace('\n', '').strip())
    for i in soup.find_all('div', class_ = 'item_body'):
        lst.append(i.text.replace('\n', '').strip())
    for i in range(0, len(lst)+1, 3):
        if i+2 > len(lst):
            break
        else:
            CTC.append(lst[i+1])
            Apply_date.append(lst[i+2])
    df = pd.DataFrame({'Job Title': job_title,
                       'Company': company,
                       'CTC': CTC,
                       'Apply Date': Apply_date})
    return df
internshala_jobs('https://internshala.com/fresher-jobs')

Job Title  \
0           Research And Communications Associate    
1                                  Data Scientist    
2                   Data Science Django Developer    
3                       Corporate Sales Associate    
4                            Full Stack Developer    
5                     Full Stack Developer (MERN)    
6                  Business Development Executive    
7                Sales Development Representative    
8                                Graphic Designer    
9                            Operations Executive    
10                 Business Development Executive    
11                    Admission Counselor (Sales)    
12         Sales And Business Development Analyst    
13                Growth Hacking Digital Marketer    
14                       Junior Node.js Developer    
15             Junior Digital Marketing Executive    
16                              iOS App Developer    
17                                Food Journalist    
18                      Junior Software Developer    
19                           Full Stack Developer    
20                               Business Analyst    
21                 Associate Full Stack Developer    
22                            Full Stack Engineer    
23                               Product Marketer    
24                 Business Development Executive    
25               Associate - Business Development    
26                   Research Analyst (Economics)    
27                      Machine Learning Engineer    
28                               Business Analyst    
29  Business Development Executive (Inside Sales)    
30                 Business Development Associate    
31                   Full Stack Software Engineer    
32                  Associate Front End Developer    
33                   Talent Acquisition Executive    
34                           Full Stack Developer    
35                              Backend Developer    
36                           Full Stack Developer    
37                         Junior Sales Associate    
38                   Digital Marketing Specialist    
39                          Android App Developer    

                                              Company            CTC  \
0       Institute For Governance, Policies & Politics    3 - 4.5 LPA   
1                                        ProtonAutoML    3 - 3.5 LPA   
2                                           Markytics    3 - 4.5 LPA   
3                                           MiM-Essay    5 - 6.5 LPA   
4                                          Wondermail      4 - 7 LPA   
5                                      Project Tinker    4 - 7.2 LPA   
6                                      Kaizen Academy    4.5 - 6 LPA   
7                                          Amigobulls  5.5 - 6.5 LPA   
8                            Underground Movement LLP      3 - 5 LPA   
9              Alphacore Technologies Private Limited    3 - 3.5 LPA   
10                 Ruhcom Enterprises Private Limited          3 LPA   
11                                Sky Education Group          3 LPA   
12                                     Mo's F&B Group    3 - 3.1 LPA   
13                                         Sleep Love    3 - 3.5 LPA   
14                                   Askadmissions.ai      3 - 4 LPA   
15                                              Krivy          3 LPA   
16               Ascentspark Software Private Limited    3.6 - 7 LPA   
17                                     Truffle Nation    3 - 4.5 LPA   
18              Habitate Technologies Private Limited    5.2 - 7 LPA   
19                                            SoluLab      3 - 5 LPA   
20                                            SoluLab    3 - 4.2 LPA   
21            REPOZITORY TECHNOLOGIES PRIVATE LIMITED    3 - 3.6 LPA   
22                              Beehive Academy India  4.2 - 8.2 LPA   
23                                                Bip      3 - 6 LPA   
24                        